File: robert-server-ws-rest/src/test/java/fr/gouv/stopc/robertserver/ws/controller/UnregisterControllerTest.java
package fr.gouv.stopc.robertserver.ws.controller;

import fr.gouv.stopc.robert.server.common.service.RobertClock;
import fr.gouv.stopc.robertserver.database.model.Registration;
import fr.gouv.stopc.robertserver.ws.test.AuthDataManager.AuthRequestData;
import fr.gouv.stopc.robertserver.ws.test.IntegrationTest;
import fr.gouv.stopc.robertserver.ws.vo.UnregisterRequestVo;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;
import org.springframework.beans.factory.annotation.Autowired;

import static fr.gouv.stopc.robertserver.ws.test.GrpcMockManager.givenCryptoServerRaiseError430ForEbid;
import static fr.gouv.stopc.robertserver.ws.test.GrpcMockManager.givenCryptoServerRaiseErrorForMacStartingWith;
import static fr.gouv.stopc.robertserver.ws.test.MockServerManager.verifyNoInteractionsWithPushNotifServer;
import static fr.gouv.stopc.robertserver.ws.test.MockServerManager.verifyPushNotifServerReceivedUnregisterForToken;
import static fr.gouv.stopc.robertserver.ws.test.MongodbManager.assertThatRegistrations;
import static fr.gouv.stopc.robertserver.ws.test.MongodbManager.givenRegistrationExistsForUser;
import static fr.gouv.stopc.robertserver.ws.test.matchers.Base64Matcher.toBase64;
import static io.restassured.RestAssured.given;
import static io.restassured.http.ContentType.JSON;
import static org.hamcrest.Matchers.*;
import static org.springframework.http.HttpStatus.*;

@IntegrationTest
class UnregisterControllerTest {

    @Autowired
    private RobertClock clock;

    @ParameterizedTest
    @MethodSource("fr.gouv.stopc.robertserver.ws.test.AuthDataManager#acceptableAuthParameters")
    void can_unregister(AuthRequestData auth) {
        givenRegistrationExistsForUser("user___1");

        given()
                .contentType(JSON)
                .body(
                        UnregisterRequestVo.builder()
                                .ebid(toBase64("user___1"))
                                .epochId(auth.epochId())
                                .time(auth.base64Time32())
                                .mac(auth.base64Mac())
                                .build()
                )

                .when()
                .post("/api/v6/unregister")

                .then()
                .statusCode(OK.value())
                .body("success", equalTo(true))
                .body("message", nullValue());

        verifyNoInteractionsWithPushNotifServer();

        assertThatRegistrations()
                .isEmpty();
    }

    @ParameterizedTest
    @MethodSource("fr.gouv.stopc.robertserver.ws.test.AuthDataManager#acceptableAuthParameters")
    void can_unregister_apple_device(AuthRequestData auth) {
        givenRegistrationExistsForUser("user___1");

        given()
                .contentType(JSON)
                .body(
                        UnregisterRequestVo.builder()
                                .ebid(toBase64("user___1"))
                                .epochId(auth.epochId())
                                .time(auth.base64Time32())
                                .mac(auth.base64Mac())
                                .pushToken("valid-token-user___1")
                                .build()
                )

                .when()
                .post("/api/v6/unregister")

                .then()
                .statusCode(OK.value())
                .body("success", equalTo(true))
                .body("message", nullValue());

        verifyPushNotifServerReceivedUnregisterForToken("valid-token-user___1");

        assertThatRegistrations()
                .isEmpty();
    }

    @ParameterizedTest
    @MethodSource("fr.gouv.stopc.robertserver.ws.test.AuthDataManager#unacceptableAuthParameters")
    void cant_unregister_with_too_much_time_drift(AuthRequestData auth) {
        givenRegistrationExistsForUser("user___1");

        given()
                .contentType(JSON)
                .body(
                        UnregisterRequestVo.builder()
                                .ebid(toBase64("user___1"))
                                .epochId(auth.epochId())
                                .time(auth.base64Time32())
                                .mac(auth.base64Mac())
                                .build()
                )

                .when()
                .post("/api/v6/unregister")

                .then()
                .statusCode(BAD_REQUEST.value())
                .body(emptyString());

        verifyNoInteractionsWithPushNotifServer();

        assertThatRegistrations()
                .extracting(Registration::getPermanentIdentifier)
                .containsExactly("idA for user___1".getBytes());
    }

    @Test
    void cant_unregister_for_a_key_unknown_by_crypto_server() {
        givenCryptoServerRaiseError430ForEbid("miss-key");

        given()
                .contentType(JSON)
                .body(
                        UnregisterRequestVo.builder()
                                .ebid(toBase64("miss-key"))
                                .epochId(clock.now().asEpochId())
                                .time(toBase64(clock.now().asTime32()))
                                .mac(toBase64("fake mac having a length of exactly 44 characters", 32))
                                .build()
                )

                .when()
                .post("/api/v6/unregister")

                .then()
                .statusCode(430)
                .body(emptyString());
    }

    @Test
    void bad_request_on_mac_validation_error_raised_by_crypto_server() {
        givenRegistrationExistsForUser("user___1");
        givenCryptoServerRaiseErrorForMacStartingWith("invalid");

        given()
                .contentType(JSON)
                .body(
                        UnregisterRequestVo.builder()
                                .ebid(toBase64("user___1"))
                                .epochId(clock.now().asEpochId())
                                .time(toBase64(clock.now().asTime32()))
                                .mac(toBase64("invalid fake mac having a length of exactly 44 characters", 32))
                                .build()
                )

                .when()
                .post("/api/v6/unregister")

                .then()
                .statusCode(BAD_REQUEST.value())
                .body(emptyString());

        assertThatRegistrations()
                .extracting(Registration::getPermanentIdentifier)
                .containsExactly("idA for user___1".getBytes());
    }

    @Test
    void not_found_for_an_unexisting_registration() {
        given()
                .contentType(JSON)
                .body(
                        UnregisterRequestVo.builder()
                                .ebid(toBase64("no-regis"))
                                .epochId(clock.now().asEpochId())
                                .time(toBase64(clock.now().asTime32()))
                                .mac(toBase64("fake mac having a length of exactly 44 characters", 32))
                                .build()
                )

                .when()
                .post("/api/v6/unregister")

                .then()
                .statusCode(NOT_FOUND.value())
                .body(emptyString());
    }

}

File: robert-server-ws-rest/src/test/kotlin/fr/gouv/stopc/robertserver/ws/controller/UnregisterControllerTest.kt
package fr.gouv.stopc.robertserver.ws.controller

import fr.gouv.stopc.robertserver.common.base64Encode
import fr.gouv.stopc.robertserver.test.MongodbManager.Companion.assertThatRegistrationCollection
import fr.gouv.stopc.robertserver.test.MongodbManager.Companion.givenRegistrationExistsForIdA
import fr.gouv.stopc.robertserver.ws.test.AuthDataManager.AuthRequestData
import fr.gouv.stopc.robertserver.ws.test.GrpcMockManager.Companion.givenCryptoServerRaiseError400ForEbid
import fr.gouv.stopc.robertserver.ws.test.GrpcMockManager.Companion.givenCryptoServerRaiseMissingDailyKeyForEbid
import fr.gouv.stopc.robertserver.ws.test.IntegrationTest
import fr.gouv.stopc.robertserver.ws.test.MockServerManager.Companion.verifyNoInteractionsWithPushNotifServer
import fr.gouv.stopc.robertserver.ws.test.When
import fr.gouv.stopc.robertserver.ws.test.matchers.verifyExceededTimeDriftIsProperlyHandled
import io.restassured.RestAssured.given
import io.restassured.http.ContentType.JSON
import org.hamcrest.Matchers.emptyString
import org.hamcrest.Matchers.equalTo
import org.hamcrest.Matchers.nullValue
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.MethodSource
import org.springframework.http.HttpStatus.BAD_REQUEST
import org.springframework.http.HttpStatus.NOT_FOUND
import org.springframework.http.HttpStatus.OK

@IntegrationTest
class UnregisterControllerTest {

    @ParameterizedTest
    @MethodSource("fr.gouv.stopc.robertserver.ws.test.AuthDataManager#acceptableAuthParameters")
    fun can_unregister(auth: AuthRequestData) {
        givenRegistrationExistsForIdA("idA_1")

        given()
            .contentType(JSON)
            .body(
                """
                    {
                      "ebid":    "${"idA_1eb1".base64Encode()}",
                      "epochId": "${auth.epochId()}",
                      "time":    "${auth.base64Time32()}",
                      "mac":     "${auth.base64Mac()}",
                      "pushInfo": ${auth.pushInfo}
                    }
                """.trimIndent()
            )
            .When()
            .post("/api/v6/unregister")
            .then()
            .statusCode(OK.value())
            .body("success", equalTo(true))
            .body("message", nullValue())

        assertThatRegistrationCollection()
            .isEmpty()
        verifyNoInteractionsWithPushNotifServer()
    }

    @ParameterizedTest
    @MethodSource("fr.gouv.stopc.robertserver.ws.test.AuthDataManager#unacceptableAuthParameters")
    fun cant_unregister_with_too_much_time_drift(auth: AuthRequestData) {
        givenRegistrationExistsForIdA("idA_1")

        given()
            .contentType(JSON)
            .body(
                """
                    {
                      "ebid":    "${"idA_1eb1".base64Encode()}",
                      "epochId": "${auth.epochId()}",
                      "time":    "${auth.base64Time32()}",
                      "mac":     "${auth.base64Mac()}",
                      "pushInfo": ${auth.pushInfo}
                    }
                """.trimIndent()
            )
            .When()
            .post("/api/v6/unregister")
            .then()
            .statusCode(BAD_REQUEST.value())
            .body(emptyString())

        verifyExceededTimeDriftIsProperlyHandled("idA_1", auth.timeDrift)
        assertThatRegistrationCollection()
            .extracting("permanentIdentifier")
            .containsExactly("idA_1")
        verifyNoInteractionsWithPushNotifServer()
    }

    @ParameterizedTest
    @MethodSource("fr.gouv.stopc.robertserver.ws.test.AuthDataManager#acceptableAuthParameters")
    fun cant_unregister_for_a_key_unknown_by_crypto_server(auth: AuthRequestData) {
        givenRegistrationExistsForIdA("idA_1")
        givenCryptoServerRaiseMissingDailyKeyForEbid("idA_1eb1")

        given()
            .contentType(JSON)
            .body(
                """
                    {
                      "ebid":    "${"idA_1eb1".base64Encode()}",
                      "epochId": "${auth.epochId()}",
                      "time":    "${auth.base64Time32()}",
                      "mac":     "${auth.base64Mac()}",
                      "pushInfo": ${auth.pushInfo}
                    }
                """.trimIndent()
            )
            .When()
            .post("/api/v6/unregister")
            .then()
            .statusCode(430)
            .body(emptyString())
    }

    @ParameterizedTest
    @MethodSource("fr.gouv.stopc.robertserver.ws.test.AuthDataManager#acceptableAuthParameters")
    fun bad_request_on_mac_validation_error_raised_by_crypto_server(auth: AuthRequestData) {
        givenRegistrationExistsForIdA("idA_1")
        givenCryptoServerRaiseError400ForEbid("idA_1eb1")

        given()
            .contentType(JSON)
            .body(
                """
                    {
                      "ebid":    "${"idA_1eb1".base64Encode()}",
                      "epochId": "${auth.epochId()}",
                      "time":    "${auth.base64Time32()}",
                      "mac":     "${auth.base64Mac()}",
                      "pushInfo": ${auth.pushInfo}
                    }
                """.trimIndent()
            )
            .When()
            .post("/api/v6/unregister")
            .then()
            .statusCode(BAD_REQUEST.value())
            .body(emptyString())

        assertThatRegistrationCollection()
            .extracting("permanentIdentifier")
            .containsExactly("idA_1")
    }

    @ParameterizedTest
    @MethodSource("fr.gouv.stopc.robertserver.ws.test.AuthDataManager#acceptableAuthParameters")
    fun not_found_for_an_unexisting_registration(auth: AuthRequestData) {
        given()
            .contentType(JSON)
            .body(
                """
                    {
                      "ebid":    "${"idA_1eb1".base64Encode()}",
                      "epochId": "${auth.epochId()}",
                      "time":    "${auth.base64Time32()}",
                      "mac":     "${auth.base64Mac()}",
                      "pushInfo": ${auth.pushInfo}
                    }
                """.trimIndent()
            )
            .When()
            .post("/api/v6/unregister")
            .then()
            .statusCode(NOT_FOUND.value())
            .body(emptyString())
    }
}
